QueryIndexMigrationTest.testIndexCleanupOnMigration

```
org.junit.runners.model.TestTimedOutException: test timed out after 60000 milliseconds
	at sun.misc.Unsafe.park(Native Method)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:182)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:822)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:980)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1292)
	at java.util.concurrent.FutureTask$Sync.innerGet(FutureTask.java:229)
	at java.util.concurrent.FutureTask.get(FutureTask.java:94)
	at com.hazelcast.map.impl.query.QueryIndexMigrationTest.testIndexCleanupOnMigration(QueryIndexMigrationTest.java:185)

```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-IbmJDK1.6/com.hazelcast$hazelcast/712/testReport/junit/com.hazelcast.map.impl.query/QueryIndexMigrationTest/testIndexCleanupOnMigration/

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OpenJDK6/com.hazelcast$hazelcast/693/testReport/junit/com.hazelcast.map.impl.query/QueryIndexMigrationTest/testIndexCleanupOnMigration/

Another kind of assertion error may be also thrown after several repetitions of the test:
```
testIndexCleanupOnMigration(com.hazelcast.map.impl.query.QueryIndexMigrationTest)  Time elapsed: 4,624.701 sec  <<< ERROR!
java.util.concurrent.ExecutionException: java.lang.AssertionError: expected:<1> but was:<2>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:631)
        at com.hazelcast.map.impl.query.QueryIndexMigrationTest.updateMapAndRunQuery(QueryIndexMigrationTest.java:253)
        at com.hazelcast.map.impl.query.QueryIndexMigrationTest.access$100(QueryIndexMigrationTest.java:64)
```

https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/3892/

java.lang.AssertionError: expected:<1> but was:<2>
Stacktrace

java.util.concurrent.ExecutionException: java.lang.AssertionError: expected:<1> but was:<2>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:645)
    at org.junit.Assert.assertEquals(Assert.java:631)
    at com.hazelcast.map.impl.query.QueryIndexMigrationTest.updateMapAndRunQuery(QueryIndexMigrationTest.java:251)
    at com.hazelcast.map.impl.query.QueryIndexMigrationTest.access$100(QueryIndexMigrationTest.java:63)```

`java.lang.AssertionError: expected:<1> but was:<2>` failure mode:

* A member executing `IMap.values(Predicate)` dispatches `QueryOperation`s on all members of the cluster
* Some members execute the query and respond with results for a set of partitions `A`
* Migrations occur
* Other members execute the query and respond with results for a set of partitions `B`
* (other `QueryPartitionOperation`s may be also executed but these are irrelevant for this case)
* Results are aggregated on the original member (https://github.com/hazelcast/hazelcast/blob/d2da655a71ba12db743066c33be55f74412c7dc2/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java#L543-L557).
* If set `A` & `B` have a non-empty intersection (due to ownership of partitions having been migrated from members who have already responded before the migration to nodes who respond after the migration), then results for some partitions will be included in the results twice.

I think that this problem started manifesting (more frequently) after #8403 was merged (which fixed a case in which query results were omitted if indexes were being rebuilt during migration at the time of query).

Considering two solutions:
* On the result aggregation side, if the results being added report a partition ID which is not found in the collection of pending partition IDs, this indicates we are about to insert twice results for a partition, so instead discard the whole result set. This is implemented in #8570.
  - Pros: simple, isolated fix. The new code will only get executed when migrations are taking place and not interfere with happy-path execution
  - Cons: when double partition result counting is detected, a whole member's results will be discarded as there is currently no way to know which result came from which partition and selectively discard only these. The missing partitions' results will be retrieved using one `QueryPartitionOperation` per partition which performs a full partition scan.
* On the member querying-side, add an `int[] partitionIds` parameter so that the member that dispatches the `QueryOperation`s asks members for results from specific partition IDs; since the requested partition IDs are determined from the partition state as observed by the dispatcher-member at the beginning of query, there can be no duplicates. Members who perform migrations in-between will report results for the intersection of (owned partition IDs, requested-for partition IDs), so in worst case a few more `QueryPartitionOperation`s will be executed to cater for results missing from specific partition IDs. This is implemented (not fully) in #8569.
  - Pros: still allows using indexes to obtain results from members, does not waste all of a members' results in case of migration
  - Cons: since indexes are not per-partition, we need to iterate over indexed query results and filter per partition ID. Affects code execution in stable cluster, not just during migrations.

```
java.util.concurrent.ExecutionException: java.lang.AssertionError: expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.map.impl.query.QueryIndexMigrationTest.updateMapAndRunQuery(QueryIndexMigrationTest.java:251)
	at com.hazelcast.map.impl.query.QueryIndexMigrationTest.access$100(QueryIndexMigrationTest.java:63)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/981/testReport/junit/com.hazelcast.map.impl.query/QueryIndexMigrationTest/testIndexCleanupOnMigration/

```
java.lang.AssertionError: expected:<30> but was:<29>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.map.impl.query.QueryIndexMigrationTest.testQueryWithIndexesWhileMigrating(QueryIndexMigrationTest.java:156)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-ZuluJDK7/com.hazelcast$hazelcast/276/testReport/junit/com.hazelcast.map.impl.query/QueryIndexMigrationTest/testQueryWithIndexesWhileMigrating/

--- a/checkstyle/suppressions.xml
+++ b/checkstyle/suppressions.xml
@@ -305,6 +305,7 @@
     <suppress checks="ClassDataAbstractionCoupling" files="com/hazelcast/map/impl/tx/TransactionalMapProxySupport"/>
     <suppress checks="NPathComplexity|CyclomaticComplexity"
               files="com/hazelcast/map/impl/querycache/event/DefaultQueryCacheEventData"/>
+    <suppress checks="MethodCount" files="com/hazelcast/map/impl/query/MapQueryEngineImpl"/>
     <suppress checks="NPathComplexity|CyclomaticComplexity" files="com/hazelcast/map/impl/SimpleEntryView"/>
 
     <!-- Util (written by Doug Lea, Agrona backport etc.) -->

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngine.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngine.java
@@ -35,6 +35,10 @@ public interface MapQueryEngine {
      * - Accepts PagingPredicate
      * - Query executed in the calling thread
      * - predicate evaluation will be parallelized if QUERY_PREDICATE_PARALLEL_EVALUATION is enabled or a PagingPredicate is used.
+     * - may return empty QueryResult (with a properly initialized result limit size) if query is executed during migrations and
+     * migration conditions do not allow for safe retrieval of results (ie. there are migrations in flight or partition state
+     * version is different at the end of query vs the version observed at the beginning of query). In this case, partitionIds
+     * are not set on the QueryResult, so callers will ignore these results and fallback to {@code QueryPartitionOperation}s.
      *
      * @param mapName   the name of the map
      * @param predicate the predicate

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java
@@ -129,10 +129,17 @@ public class MapQueryEngineImpl implements MapQueryEngine {
         QueryResult result = tryQueryUsingIndexes(predicate, initialPartitions, mapContainer, iterationType,
                 initialPartitionStateVersion);
         if (result == null) {
-            result = queryUsingFullTableScan(mapName, predicate, initialPartitions, iterationType);
+            result = querySafelyUsingFullTableScan(mapName, predicate, initialPartitions, iterationType,
+                    initialPartitionStateVersion);
         }
 
-        if (hasPartitionVersion(initialPartitionStateVersion, predicate)) {
+        if (result == null) {
+            // if fallback to full table scan also failed to return any results due to migrations,
+            // then return empty result set without any partition IDs set (so that it is ignored by callers).
+            result = newQueryResult(initialPartitions.size(), iterationType);
+        } else if (hasPartitionVersion(initialPartitionStateVersion, predicate)) {
+            // if results have been returned and partition state version has not changed, set the partition IDs
+            // so that caller is aware of partitions from which results were obtained.
             result.setPartitionIds(initialPartitions);
         }
 
@@ -148,7 +155,8 @@ public class MapQueryEngineImpl implements MapQueryEngine {
         // MapService.getMigrationsInFlight() returns the number of currently executing migrations (for which
         // beforeMigration has been executed but commit/rollback is not yet executed).
         // This is a temporary fix for 3.7, the actual issue will be addressed with an additional migration hook in 3.8.
-        if (mapServiceContext.getService().getOwnerMigrationsInFlight() > 0) {
+        // see https://github.com/hazelcast/hazelcast/issues/6471 & https://github.com/hazelcast/hazelcast/issues/8046
+        if (hasOwnerMigrationsInFlight()) {
             return null;
         }
 
@@ -158,19 +166,64 @@ public class MapQueryEngineImpl implements MapQueryEngine {
         }
 
         QueryResult result = newQueryResult(partitions.size(), iterationType);
-        // if partition state version has changed in the meanwhile, this means migrations were executed and we may
-        // return stale data, so we should rather return null and let the query run with a full table scan
-        if (initialPartitionStateVersion != partitionService.getPartitionStateVersion()) {
+        // If partition state version has changed in the meanwhile, this means migrations were executed and we may
+        // return stale data, so we should rather return null and let the query run with a full table scan.
+        // Also make sure there are no long migrations in flight which may have started after starting the query
+        // but not completed yet.
+        if (isResultSafe(initialPartitionStateVersion)) {
+            result.addAll(entries);
+            return result;
+        } else {
             return null;
         }
-        result.addAll(entries);
-        return result;
     }
 
     protected void updateStatistics(MapContainer mapContainer) {
         if (mapContainer.getMapConfig().isStatisticsEnabled()) {
             LocalMapStatsImpl localStats = localMapStatsProvider.getLocalMapStatsImpl(mapContainer.getName());
             localStats.incrementOtherOperations();
+        }
+    }
+
+    /**
+     * Wraps {@link #queryUsingFullTableScan(String, Predicate, Collection, IterationType)} to avoid returning potentially
+     * flawed results.
+     * <ul>
+     *     <li>if owner partition migrations are executing before running the query, then return immediately {@code null}, as
+     *     results cannot be reliably obtained</li>
+     *     <li>execute the query using full table scan</li>
+     *     <li>if after the query has been executed migrations are in-flight or partition state version has changed, then
+     *     results are not considered reliable</li>
+     * </ul>
+     * @param name
+     * @param predicate
+     * @param partitions
+     * @param iterationType
+     * @param initialPartitionStateVersion
+     * @return query results or {@code null} if results are considered potentially flawed.
+     * @throws InterruptedException
+     * @throws ExecutionException
+     */
+    protected QueryResult querySafelyUsingFullTableScan(String name, Predicate predicate, Collection<Integer> partitions,
+                                                        IterationType iterationType, int initialPartitionStateVersion)
+            throws InterruptedException, ExecutionException {
+
+        QueryResult result;
+
+        if (hasOwnerMigrationsInFlight()) {
+            return null;
+        }
+
+        result = queryUsingFullTableScan(name, predicate, partitions, iterationType);
+
+        // If partition state version has changed in the meanwhile, this means migrations were executed and we may
+        // return stale data, so we should rather return null.
+        // Also make sure there are no long migrations in flight which may have started after starting the query
+        // but not completed yet.
+        if (isResultSafe(initialPartitionStateVersion)) {
+            return result;
+        } else {
+            return null;
         }
     }
 
@@ -264,7 +317,7 @@ public class MapQueryEngineImpl implements MapQueryEngine {
     }
 
     protected boolean hasPartitionVersion(int expectedVersion, Predicate predicate) {
-        if (expectedVersion != partitionService.getPartitionStateVersion()) {
+        if (hasPartitionStateVersionChanged(expectedVersion)) {
             logger.info("Partition assignments changed while executing query: " + predicate);
             return false;
         }
@@ -609,5 +662,44 @@ public class MapQueryEngineImpl implements MapQueryEngine {
             MapQueryEngineImpl queryEngine = (MapQueryEngineImpl) mapServiceContext.getMapQueryEngine(name);
             return queryEngine.queryTheLocalPartition(name, predicate, partition);
         }
+    }
+
+    /**
+     * Check whether migrations of owner partition are currently executed.
+     * If a migration is in progress, do not attempt to use an index as they may have not been created yet.
+     * MapService.getMigrationsInFlight() returns the number of currently executing migrations (for which
+     * beforeMigration has been executed but commit/rollback is not yet executed).
+     * This check is a temporary fix for 3.7, the actual issue will be addressed with an additional migration hook in 3.8.
+     * see https://github.com/hazelcast/hazelcast/issues/6471 & https://github.com/hazelcast/hazelcast/issues/8046
+     * @return {@code true} if owner partition migrations are currently being executed, otherwise false.
+     * @see com.hazelcast.spi.impl.DelegatingMigrationAwareService
+     */
+    protected boolean hasOwnerMigrationsInFlight() {
+        return mapServiceContext.getService().getOwnerMigrationsInFlight() > 0;
+    }
+
+    /**
+     * Check whether partition state version has changed since {@code initialPartitionStateVersion}.
+     * @param initialPartitionStateVersion the initial partition state version to compare against
+     * @return {@code true} if current partition state version is not equal to {@code initialPartitionStateVersion}
+     */
+    protected boolean hasPartitionStateVersionChanged(int initialPartitionStateVersion) {
+        return initialPartitionStateVersion != partitionService.getPartitionStateVersion();
+    }
+
+    /**
+     * Check whether results obtained since partition state version was at {@code initialPartitionStateVersion} are safe to be
+     * returned to the caller. Effectively this method checks:
+     * <ul>
+     *     <li>whether owner migrations are currently in flight; if there are any, then results are considered flawed</li>
+     *     <li>whether current partition state version has changed, implying that some migrations were executed in the
+     *     meanwhile, so results are again considered flawed</li>
+     * </ul>
+     * @param initialPartitionStateVersion
+     * @return {@code true} if no owner migrations are currently executing and {@code initialPartitionStateVersion} is the same
+     * as the current partition state version, otherwise {@code false}.
+     */
+    protected boolean isResultSafe(int initialPartitionStateVersion) {
+        return !hasOwnerMigrationsInFlight() && !hasPartitionStateVersionChanged(initialPartitionStateVersion);
     }
 }

